indico/modules/rb_new/client/js/components/RoomListFilters.jsx

```diff
@@ -0,0 +1,182 @@
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Indico; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+import propTypes from 'prop-types';
+import React from 'react';
+import {Button, Form, Icon, Input, Popup, Select} from 'semantic-ui-react';
+import {DebounceInput} from 'react-debounce-input';
+
+import {Translate} from 'indico/react/i18n';
+import {parseRoomListFiltersText} from '../util';
+
+import './RoomListFilters.module.scss';
+
+
+export default class RoomListFilters extends React.Component {
+    constructor(props) {
+        super(props);
+        this.state = {
+            filtersVisible: false,
+            text: '',
+            building: '',
+            floor: ''
+        };
+    }
+
+    componentDidMount() {
+        const {fetchBuildings} = this.props;
+        fetchBuildings();
+    }
+
+    handleVisibleChange(visible) {
+        this.setState({filtersVisible: visible});
+    }
+
+    applyFilters() {
+        const {onConfirm} = this.props;
+        this.setState({filtersVisible: false});
+        onConfirm();
+    }
+
+    handleFiltersChange(filter, value) {
+        const {setTextParamFilter, setAdvancedParamFilter, filters: {text}} = this.props;
+        const stateUpdates = {[filter]: value};
+
+        if (filter === 'building') {
+            stateUpdates.floor = '';
+        } else if (filter === 'text') {
+            const parsedValues = parseRoomListFiltersText(value);
+
+            stateUpdates.building = parsedValues.building;
+            stateUpdates.floor = parsedValues.floor;
+        }
+
+        this.setState(stateUpdates, () => {
+            let textValue = '';
+            const {building, floor} = this.state;
+
+            if (filter === 'text') {
+                textValue = value;
+            } else {
+                if (building) {
+                    textValue = `building:${building}`;
+                }
+
+                if (floor) {
+                    textValue += ` floor:${floor}`;
+                }
+            }
+
+            this.setState({text: textValue});
+            setAdvancedParamFilter('text', textValue);
+        });
+
+        if (filter === 'text') {
+            if (!text || value.trim() !== text.trim()) {
+                setTextParamFilter(value);
+            }
+        }
+    }
+
+    render() {
+        const commonAttrs = {
+            search: true,
+            selection: true,
+            style: {width: 200}
+        };
+
+        const {buildings: {list: buildingsList, isFetching}} = this.props;
+        const {building, floor} = this.state;
+        let floors = [{text: '', value: ''}];
+
+        if (building && buildingsList[building]) {
+            floors = floors.concat(buildingsList[building].floors.map((floorItem) => ({
```

Comment:
I'd assign this to `buildingFloors` and later use `[...floors, ...buildingFloors]` to merge the two